src/mscorlib/src/System/Enum.cs

```diff
@@ -727,6 +709,94 @@ internal unsafe Object GetValue()
             }
         }
 
+        [System.Security.SecuritySafeCritical]
+        private unsafe ulong GetValueAsULong()
+        {
+            fixed (void* pValue = &JitHelpers.GetPinningHelper(this).m_data)
+            {
+                switch (InternalGetCorElementType())
+                {
+                    case CorElementType.I1:
+                        return (ulong)*(sbyte*)pValue;
+                    case CorElementType.U1:
+                        return *(byte*)pValue;
+                    case CorElementType.Boolean:
+                        return Convert.ToUInt64(*(bool*)pValue, CultureInfo.InvariantCulture);
+                    case CorElementType.I2:
+                        return (ulong)*(short*)pValue;
+                    case CorElementType.U2:
+                    case CorElementType.Char:
+                        return *(ushort*)pValue;
+                    case CorElementType.I4:
+                        return (ulong)*(int*)pValue;
+                    case CorElementType.U4:
+                    case CorElementType.R4:
+                        return *(uint*)pValue;
+                    case CorElementType.I8:
+                        return (ulong)*(long*)pValue;
+                    case CorElementType.U8:
+                    case CorElementType.R8:
+                        return *(ulong*)pValue;
+                    case CorElementType.I:
+                        if (IntPtr.Size == 8)
+                        {
+                            return *(ulong*)pValue;
+                        }
+                        else
+                        {
+                            return (ulong)*(int*)pValue;
+                        }
+                    case CorElementType.U:
+                        if (IntPtr.Size == 8)
+                        {
+                            return *(ulong*)pValue;
+                        }
+                        else
+                        {
+                            return *(uint*)pValue;
+                        }
+                    default:
+                        Contract.Assert(false, "Invalid primitive type");
+                        return 0;
+                }
+            }
+        }
+
+        [System.Security.SecuritySafeCritical]
+        private unsafe static ulong GetValueAsULong(object value)
```

Comment:
I agree that `ToUInt64(object)` is not very efficient. I think it should be possible to merge `ToUInt64` and `GetValueAsULong(object)` so that there is just one method that does it efficiently.